pkg/storage/fs/ocis/ocis.go

```diff
@@ -207,8 +195,83 @@ func (fs *ocisfs) CreateDir(ctx context.Context, fn string) (err error) {
 	return fs.tp.CreateDir(ctx, node)
 }
 
-func (fs *ocisfs) CreateReference(ctx context.Context, path string, targetURI *url.URL) error {
-	return fs.tp.CreateReference(ctx, path, targetURI)
+// The share folder, aka "/Shares", is a virtual thing.
+func (fs *ocisfs) isShareFolder(ctx context.Context, p string) bool {
+	// TODO what about a folder '/Sharesabc' ... would also match
+	return strings.HasPrefix(p, fs.pw.ShareFolder)
+}
+
+func (fs *ocisfs) isShareFolderRoot(ctx context.Context, p string) bool {
+	return path.Clean(p) == fs.pw.ShareFolder
+}
+
+func (fs *ocisfs) isShareFolderChild(ctx context.Context, p string) bool {
+	p = path.Clean(p)
+	vals := strings.Split(p, fs.pw.ShareFolder+"/")
+	return len(vals) > 1 && vals[1] != ""
+}
+func (fs *ocisfs) getInternalHome(ctx context.Context) (string, error) {
+	if !fs.pw.EnableHome {
+		return "", errtypes.NotSupported("ocisfs: get home not enabled")
+	}
+
+	u, ok := user.ContextGetUser(ctx)
+	if !ok {
+		return "", errtypes.InternalError("ocisfs: wrap: no user in ctx and home is enabled")
+	}
+
+	relativeHome := templates.WithUser(u, fs.pw.UserLayout)
+	return relativeHome, nil
+}
+
+// TODO there really is no difference between the /Shares folder and normal nodes
+// we can store it is a node
+// but when home is enabled the "/Shares" folder should not be listed ...
+// so when listing home we need to check that flag and add the shared node
+
+// /Shares/mounted1
+// /Shares/mounted1/foo/bar
+// references should go into a Shares folder
+func (fs *ocisfs) CreateReference(ctx context.Context, p string, targetURI *url.URL) (err error) {
+
+	p = strings.Trim(p, "/")
+	parts := strings.Split(p, "/")
+
+	if len(parts) != 2 {
+		return errtypes.PermissionDenied("ocisfs: references must be a child of the share folder: share_folder=" + fs.pw.ShareFolder + " path=" + p)
+	}
+
+	if parts[0] != strings.Trim(fs.pw.ShareFolder, "/") {
+		return errtypes.PermissionDenied("ocisfs: cannot create references outside the share folder: share_folder=" + fs.pw.ShareFolder + " path=" + p)
+	}
+
+	// create Shares folder if it does not exist
+	var n *Node
+	if n, err = fs.pw.NodeFromPath(ctx, fs.pw.ShareFolder); err != nil {
+		return errtypes.InternalError(err.Error())
+	} else if !n.Exists {
+		if err = fs.tp.CreateDir(ctx, n); err != nil {
+			return
+		}
+	}
+
+	if n, err = n.Child(parts[1]); err != nil {
+		return errtypes.InternalError(err.Error())
+	}
+
+	if n.Exists {
+		return errtypes.AlreadyExists(p)
+	}
+
+	if err = fs.tp.CreateDir(ctx, n); err != nil {
+		return
+	}
+
+	internal := filepath.Join(fs.pw.Root, "nodes", n.ID)
+	if err = xattr.Set(internal, referenceAttr, []byte(targetURI.String())); err != nil {
```

Comment:
I guess we have no way of validating if the reference is valid or at least has the correct format (if it's a URL)

Reply:
we have, we could and we should so that we can use a symbolic link if the cs3 reference uses a node in the same storage provider. but that would only allow the admin to navigate the reference in that case ... not important right now.